pages/api/companies.ts

```diff
@@ -95,31 +102,77 @@ async function handlePost(userId: string, req: NextApiRequest, res: NextApiRespo
     companyUrl: removeProtocolAndWwwIfPresent(req.body.companyUrl),
   };
 
-  const duplicateCompany = await prisma.company.findFirst({
-    where: {
-      name: companyPostData.name,
-      companyUrl: companyPostData.companyUrl,
-    },
-    select: { id: true, name: true, companyUrl: true },
-  });
+  const duplicateCompany = await findCompany(companyPostData.name, companyPostData.companyUrl);
+
+  const isNewCompany = duplicateCompany === null;
+  if (isNewCompany) {
+    const createdCompany = await createCompany(userId, companyPostData);
+
+    res
+      .status(HttpStatus.CREATED)
+      .json(createJsonResponse(createdCompany, messages[MessageType.COMPANY_CREATED_SUCCESSFULLY]));
+    return;
+  }
 
-  if (duplicateCompany) {
+  const isAlreadyContributedByUser = duplicateCompany.contributions.some(
+    (contribution) => contribution.contributorId === userId,
+  );
+
+  if (duplicateCompany.isVerified || isAlreadyContributedByUser) {
     res.status(HttpStatus.OK).json(createJsonResponse(duplicateCompany, messages[MessageType.COMPANY_ALREADY_EXISTS]));
```

Comment:
do we need to update the type `CompanyData` if we pass in the entire `duplicateCompany`? But `createdCompany` doesn't have the extra fields. I think we can take only the fields needed from `duplicateCompany`.